### PR TITLE
REST API cleanup, part 2

### PR DESCRIPTION
**Pull request by mmedenjak**
- all HTTP status codes, including HTTP 200 OK can now return a response
body
- moved helper methods from HttpPostCommandProcessor to HttpCommand
- removed the exception handling and authentication helper methods. The
exception handling is now done on one place and not by each handler
independently. The authentication helper method wasn't that great since
it required constructing lambdas and required indentation of the handler
code which was not worth the effort
- the exception handling now always returns a HTTP 500 for an error with
a JSON body containing the exception message
- the authentication and parameter parsing is now handled by the method
decodeParamsAndAuthenticate. This method will also expect a number of
parameters. If the request does not contain at least that many
parameters, HTTP 400 is returned with a JSON describing the requested
number of parameters
- HTTP 400 is NOT returned now if any handler throws an
IndexOutOfBoundsException. The assumption was that this was probably
caused by a bad request but that doesn't need to be the case.
- added credentials checking to POST methods for map and queue. This is
debatable but I assumed that mutation methods should be authenticated
(the DELETE method isn't but it requires a bigger change to add request
data to HttpDeleteCommand)
- added some javadoc, Nonnull and Nullable annotations and some minor
cleanups

Possible future changes:
- add parameter validations and return HTTP 400 on bad parameters
- return HTTP 503 or HTTP 501 when EE handlers are used on OS side
instead of returning HTTP 200 or HTTP 500 (examples - setting license
key, WAN REST endpoints)

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java [hunk 1]
```diff
@@ -83,24 +102,167 @@
...
+     *                                      named character encoding is not supported
+     */
+    private static String[] decodeParams(HttpPostCommand command, int expectedParamCount)
+            throws UnsupportedEncodingException {
+        final byte[] data = command.getData();
+        final String[] encoded = bytesToString(data).split("&");
+        final String[] decoded = new String[encoded.length];
+
+        if (encoded.length < expectedParamCount) {
+            return null;
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java]: I wonder why you decided to handle error by returning `null`, not by throwing an exception. The disadvantage of this approach here is that if you forget to handle it anywhere, then you'll get `NullPointerException`.

File: hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java [hunk 2]
```diff
@@ -234,17 +233,10 @@ private void registerExecutor() {
...
 
-    public byte[] clusterWideUpdateManagementCenterUrl(String newUrl) {
-        try {
-            final Collection<Member> memberList = instance.node.clusterService.getMembers();
-            for (Member member : memberList) {
-                send(member.getAddress(), new UpdateManagementCenterUrlOperation(newUrl));
-            }
-
-            return HttpCommand.RES_204;
-        } catch (Throwable throwable) {
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java]: The exception will be caught by the `HttpPostCommandProcessor#handle` method, logged and translated into HTTP 500 with an error JSON.

File: hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java]: Right, I'll have to introduce some custom exception as not to get HTTP 500 from the post command handler. Would that be ok?

Comment (leszko) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java]: Sounds good!

File: hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java]: Seems this URI and handler were removed in a [recent commit](https://github.com/hazelcast/hazelcast/pull/16161/files#diff-3850ee7b4bd9df3672e6c4b18d1ec70cL241-L253).

File: hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpCommandProcessor.java]: Introduced exceptions for bad request and forbidden. Looks better now, thanks for the idea!

**Comment by mmedenjak** [comment]
@leszko addressed review comments in a separate PR. Seems there were _some_ tests, just that they have failed and I haven't noticed.

Changes in the new commit:
• added `HttpBadRequestException` and `HttpForbiddenException` and throwing these from the parameter parsing method, catching them in `HttpPostCommandProcessor#handle` and sending responses
• added test to test both of these for all POST URI's that require authentication. As a consequence, removed some existing (now duplicate) tests that were testing for FORBIDDEN. I didn't add tests to assert behaviour of correct (authenticated) requests as I assume there are existing tests for that.
• added more assertions to assert JSON output on some POST tests that only asserted http status code
• removed authentication from map/queue POST URIs. It wasn't there before this PR and introducing it also brought some issues since the handler methods expect that there are no String parameters but instead use the entire request data as a value for the map/queue. Introducing authentication as parameters complicates this. If we really want to authenticate these two, we can do it in a follow-up PR.

@emre-aydin if you want, you can re-review the PR as well, there were some non-trivial changes, as you can see above ☝️

**Comment by mmedenjak** [comment]
Failed with unrelated test failure: https://github.com/hazelcast/hazelcast/issues/16099

**Comment by mmedenjak** [comment]
Thank you for the reviews, everyone 🙇

**Pull request merged**